### DISCUSSION TITLE
Limiting versions of python used while building a formula

### DISCUSSION THREAD

**Discussion by rgoldberg** [Homebrew/discussions#5662] [Everyday usage] [answered]
### Output of `brew config`

```shell
n/a
```

### Output of `brew doctor`

```shell
n/a
```

### Description of issue

Various formulae have build dependencies on multiple versions of python (e.g., , `cryptography` depends on `python@3.11`, `python@3.12`, and `python@3.13`).

When installing/upgrading, can I limit the build / build dependencies to just one of them?

If so, must I manually specify what python versions to use/skip? Or can I just say: only build for the python versions that are currently installed?

All of my installed python formulae have `python@3.12` as their runtime dependency.

At the least, I'd prefer to be able to ignore 3.11 (i.e. not install 3.11 for build, and not output 3.11 versions of the formulae), because I presumably will never use 3.11 again.

**Comment by SMillerDev** [comment]
No, this is not possible. Builds will only every build what is defined in the formula, there is no method to modify this.

**Reply by rgoldberg** [reply]
Thanks for the info. Do you think it would be difficult to add in a flag to `brew install` that makes it ignore certain dependencies & to add support in formulae to ignore a block if a dependency was ignored? Presumably, this info would have to be remembered for subsequent `brew upgrades` to follow suit.

**Reply by p-linnane** [reply, marked as answer]
You can create your own tap and do whatever you want, but we wouldn't support something like this.

**Reply by cho-m** [reply]
For `cryptography` specifically, one thing that may be possible is updating formula to build a single copy of `cryptography` and symlinking it since it builds in limited-API / ABI-stable mode.

However, this is only possible for Python packages that support this. For example, `cryptography`'s dependency `cffi` doesn't (you can see that official wheels are Python minor version specific - https://pypi.org/project/cffi/#files). So, would still need multiple Pythons if building `cryptography` on a fresh system in order to build `cffi`.

Also, Python 3.11 usage should be removed as part of current Python 3.13 migration. Just need 0 dependents usage. Same thing as last migration - https://github.com/Homebrew/homebrew-core/pull/156970

**Reply by rgoldberg** [reply]
@cho-m Will brew ever require that all formulae that are written in Python use something like pipx, so there won't be inter-formula dependencies like awscli on cryptography (which would also remove my need for cffi)?

**Reply by cho-m** [reply]
No, some formulae will remain as dependencies for documented reasons[^1]. We use `cryptography` formula to help prevent `rust` propagation and also to get latest security updates.

[^1]: https://docs.brew.sh/Python-for-Formula-Authors#examples-of-allowed-libraries-in-homebrew-core

**Reply by rgoldberg** [reply]
@cho-m I don't have experience with directly using Rust.

Brew installs various different languages & build systems to build formulae, so just wondering: why is Rust of singular concern? Are there many incompatible versions of Rust, so, compared to 3 installs of Python for builds, there'd be like 30 installs of Rust for builds? Or is it something else?

Thanks.

**Comment by quackerex** [comment]
Maybe you can use `--ignore-dependencies`
